### PR TITLE
10 percent time/finish ab testing

### PR DESCRIPTION
**Pull request by tlehman**
This finishes the A/B testing feature in Blackbeard. Currently, we use features set at some percentage (10%, 50%, etc) and try to do ad-hoc A/B testing that way, but it's wrong for at least these reasons:
- feature view counts ≠ unique user counts
- just looking at ratios of successes to failures are unlikely to separate noise from real signal
- statistics are hard, need theory and tests, which we currently don't have in blackbeard

Features in blackbeard should be used for rollout, and Tests should be used for A/B testing. I am planning on doing a lunch-and-learn about A/B testing, the different methodologies, the pitfalls, and how to do it right.

This pull request is a spike that provides the minimum working implementation of [Bayesian A/B testing](http://www.evanmiller.org/bayesian-ab-testing.html), the **TL;DR** of why Bayesian is this: **it doesn't require you specify your experiment size in advance**, so it can be run for as long or as short as desired, and the probability that B is better than A is updated for every user or visitor that sees the test.

## Screenshot

screen shot 2017-03-21 at 9 44 56 am

### Defining an A/B test

``` ruby
@emoji = $pirate.ab_test(:emoji_test, a: "&#128568;", b: "&#128526;")

# @emoji has one of the values passed in the hash
```

### Marking a user/visitor as having "finished" the test

``` ruby
$pirate.ab_finished(:emoji_test)
```

### CODE REVIEW

File: lib/blackbeard/math.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  module Math
+    def self.beta(x,y)
+      lx = ::Math.lgamma(x).first
+      ly = ::Math.lgamma(y).first
+      lxplusy = ::Math.lgamma(x + y).first
+
+      precision = 5
+      ::BigMath.exp(lx + ly - lxplusy, precision)
+    end
+  end
```
Comment (tlehman) [inline lib/blackbeard/math.rb]: Explanation for why this function exists is [here](https://github.com/goldstar/blackbeard/blob/9af9ff46e841ae82d2819db8a02b30b738a58d93/docs/beta.org)

**Comment by rgraff** [comment]
I did a quick scan and it looks pretty good. It's a good step forward.

One of my goals for AB tests in blackbeard is to support more than 2 variations and do true multi-variate (sets of variations)--for example Headline B with Button C is the best combination. That gets harder.

**Comment by tlehman** [comment]
@rgraff Yeah, doing more than two variations is covered in the same Evan Miller article on Bayesian A/B testing, implementing it would be a bit more involved, but doable. It's always possible to set up n tests and then get 2ⁿ combinations, since we have true independence in this implementation.

Either way, I'd like this PR and the coming lunch & learn (April 6) to be a place where the backend, frontend and app teams can all discuss the pros/cons of various tools and decide together how we should do it. I've heard sixpack is an option. It sounds like a solid choice, but if we could do A/B testing without adding another endpoint, that has advantages too.

**Comment by tlehman** [comment]
We are probably going with sixpack, @deleteme and I will set up an instance for a 10% time project and try it out.

**Pull request closed without merging**